### ISSUE TITLE
Improve / document handling of authentication expiry and asynchronous permission denial

### ISSUE THREAD

**Issue by jwngr** [firebase/angularfire#514]
`$$error` should allow you to catch this, but it is not well documented. We may also want to consider adding something to `$firebaseAuth` to handle these events asynchronously.

**Comment by jwngr** [comment]
Action items:
1. Add `$onCancel()` to `$firebaseAuth` which will run a user-supplied callback when authentication expires.
2. Add `$expiresAt()` to `$firebaseAuth` to determine when authentication will expire.
3. Re-attache listeners for `$FirebaseArray` and `$FirebaseObject` once authentication returns (one time).
4. Improve documentation and add an example to the documentation on how to use `$$error`.

**Comment by wbeange** [comment]
Hey, right now I'm trying to solve how to handle JWT token expiration on my Angular client (if I can catch it, I'll call my Rails server and get a new token).

FIREBASE WARNING: auth() was canceled: Auth token is expired.

I can't find any documentation for how to handle this event. Any input is appreciated.

**Comment by jwngr** [comment]
This GitHub issue exists for the exact reason to help you accomplish what you want to accomplish :) The short answer is that we don't really support this that well right now.

For the time being, you will have to use `$extendFactory()` and put your logic in the `$$error()`method. [Here are the slim docs](https://www.firebase.com/docs/web/libraries/angular/api.html#angularfire-extending-the-factories-firebaseobjectextendfactory) on that. The `$$error()` method will fire once authentication fails due to token expiry or security rules.

You should see improvements in this area in the coming weeks and this is tagged for the 1.0.0 release.

**Comment by gigablox** [comment]
Wondering if this will help ease a `permission_denied` error whenever `$unauth()` is invoked from an authenticated state? Trying to figure out how to get rid of the error.

**Comment by jwngr** [comment]
@gigablox - I'm not really sure what you mean. It will help you catch that error, yes. But that is still an error state which will need to be handled. There is no working around that.

**Comment by katowulf** [comment]
Hi @gigablox, it won't allow you to sync while not authenticated (assuming security rules prevent this) but  we are going to try and re-establish the listeners if authentication is restored later. It's also not going to prevent the permission_denied error; you would need to--as we discussed before--call `$destroy` before `$unauth()` if you don't want to see the (benign, btw) error message in your logs.

**Comment by gigablox** [comment]
Thanks guys, just trying to understand. I have a single object I'm playing with here:

    $scope.currentUser.$destroy();
    auth.$unauth();

Still throws an error - I must be doing something wrong?

**Comment by wbeange** [comment]
http://stackoverflow.com/questions/28101432/how-do-i-register-a-jwt-token-expired-event-with-angularfire-angularfire-custo

**Comment by jwngr** [comment]
@wbeange - *Please do not cross-post questions.* You will get responses from the same exact people on StackOverflow as you do from right here (namely, Kato and myself). You are just making our lives more difficult by cross-posting and adding undue work for us and for those who are looking for answers. We have even been very responsive on this thread already, so let's please just keep the discussion here.

As for your question, I don't really know what more you want me to add. This very issue is open to resolve the exact problem you are running into. The reason it is a GitHub issue is because it is currently broken/not good enough. We are working on it and we will get it fixed as soon as we can. We appreciate your patience.

**Comment by gigablox** [comment]
Thanks for your help guys.

@katowulf Ok this makes sense now. Haven't used AngularFire/Firebase since 0.7.1. I see you folks have been busy... I'm getting started on a new project and just getting up to speed on AngularFire/Firebase API changes.

After crawling some of your responses to questions similar to my own I see what you mean by using a service to collect references. I could push on any of the authentication dependent references and iterate `obj[i].$destroy()` them before `auth.$unauth()`.

I also see what you mean by re-establishing listeners if a connection to Firebase is lost. Really interesting stuff. Is that feature implemented already or on the road-map?

**Comment by wbeange** [comment]
Thanks for the very quick feedback you guys are great!

**Comment by jwngr** [comment]
A couple notes on our proposed plan:
1. Add $onCancel() to $firebaseAuth which will run a user-supplied callback when authentication expires
2. Many of our auth tokens do not expire

**Comment by katowulf** [comment]
Note that this won't make it into the 1.0 release. We wanted to push this in but it's a complex problem and we want to get our 1.0 release out without holding it up for this issue. We're moving this to the next milestone. I'll be adding some better documentation around $$error as an intermediate step.

**Comment by birdcw** [comment]
It's my understanding that there currently isn't a callback available for when authentication is about to expire, there's only one for when it has just expired? I think it would be useful to have a callback that allows you to "clean up" while still authenticated, just before authentication expires. For example, if you're using authentication to allow users to complete a task within a certain timelimit, you could use firebase expiresAt to set the timelimit, and reject/clean up data from an incomplete task just before the expiration.

**Comment by jwngr** [comment]
Thanks for the feedback @birdcw. There are several reasons we don't provide a "your token is about to expire" event, the main one being we don't store expiry information in all tokens. So we don't know from the client-side when your token actually is going to expire. Not until we hit the server do we determine that. But we definitely will take this feedback under consideration as the authentication API evolves.

**Comment by ereyes01** [comment]
Hi folks,

I'm running into a problem whose underlying root cause might be related to this one. I have several routes whose resolve methods use the $firebaseAuth object's $requireAuth promise to ensure you are logged in before giving you access to the page. On failure, the user is redirected to the login page. I am using email login (I believe it was formerly simplelogin) to authenticate users.

The problem happens when the browser remains open for a long time and $firebaseAuth's underlying JWT expires. When I navigate back to an authenticated route, $requireAuth's promise at first succeeds, allowing me access to the authenticated route. After a few moments, if I try another authenticated route, it successfully detects the expired token and redirects me to the login screen.

On the surface and without diving into your code, this sounds like a race condition in $firebaseAuth's underlying global state and expiring tokens. Will this problem go away when this issue is fixed? Also, what is a reasonable work-around for route authentication in the meantime? I would prefer some way to hasten de-authentication over $extending the $$error of each of my Firebase objects/arrays.

Thanks, and sorry if this is better suited for a new issue. Let me know if you want me to post this in a new issue.

Thanks, Eddy

**Comment by katowulf** [comment]
@ereyes01 this deserves its own issue, while related in concept, it can be handled and implemented separately.

**Comment by ereyes01** [comment]
Thanks, opened #590

**Comment by katowulf** [comment]
I've decided not to add this in 1.0. While it's inconvenient, permission denied errors should mostly be considered development bugs. You can certainly make arguments to the contrary, and there are edge cases to handle here. However, I'd rather focus on 2.0 and ensure we handle use cases like this more elegantly there. I'm going to declare the current handling sufficient and close this as wont-fix.

Untested but confident workaround example:

```
app.factory('renewableArray', function($firebaseArray) {
  return function(ref, callback) {
    function renew(authData) {
      if( authData !== null ) {
        ref.offAuth(renew);
        callback(new Factory(ref));
      }
    }

    var Factory = $firebaseArray.$extend({
      $$error: function(err) {
        $firebaseArray.prototype.$$error.apply(this, arguments);
        if( ref.getAuth() === null ) {
          // if auth is lost, then wait for reauth and recreate
          ref.onAuth(renew);
        }
      }
    });

    // wait for auth
    ref.onAuth(renew);
  };
});

app.controller('ctrl', function(renewableArray) {
  var ref = new Firebase(...);

  renewableArray(ref, function(arr) {
    $scope.arr = arr;
  });
});
```
